File: pregnancy_weight.js
// pregnancy_weight.js
#!/usr/bin/env node
const fs = require('fs');
const { program } = require('commander');

const DATA_FILE = 'pregnancy_weight.json';

class PregnancyWeight {
    constructor() {
        this.data = this.load();
        this.height = this.data.height || null;
        this.preWeight = this.data.preWeight || null;
        this.entries = this.data.entries || [];
    }

    load() {
        if (fs.existsSync(DATA_FILE)) {
            try {
                return JSON.parse(fs.readFileSync(DATA_FILE));
            } catch (e) {}
        }
        return { height: null, preWeight: null, entries: [] };
    }

    save() {
        fs.writeFileSync(DATA_FILE, JSON.stringify({
            height: this.height,
            preWeight: this.preWeight,
            entries: this.entries
        }, null, 2));
    }

    setHeight(cm) {
        this.height = cm;
        this.save();
        console.log(`✅ Height set to ${cm} cm`);
    }

    setPreWeight(kg) {
        this.preWeight = kg;
        this.save();
        console.log(`✅ Pre‑pregnancy weight set to ${kg} kg`);
    }

    bmi(weight) {
        if (!this.height) return null;
        const h = this.height / 100;
        return weight / (h * h);
    }

    bmiCategory(bmiVal) {
        if (bmiVal < 18.5) return 'Underweight';
        if (bmiVal < 25.0) return 'Normal';
        if (bmiVal < 30.0) return 'Overweight';
        return 'Obese';
    }

    add(weight, dateStr, note) {
        if (!this.height) {
            console.log('❌ Please set your height first: height <cm>');
            return;
        }
        if (!dateStr) dateStr = new Date().toISOString().slice(0,10);
        const bmiVal = this.bmi(weight);
        const entry = { date: dateStr, weight, note: note || '', bmi: bmiVal };
        this.entries.push(entry);
        this.entries.sort((a, b) => a.date.localeCompare(b.date));
        this.save();
        console.log(`✅ Logged ${weight} kg on ${dateStr}`);
    }

    list() {
        if (this.entries.length === 0) {
            console.log('No entries.');
            return;
        }
        console.log('\n📋 Weight Entries:');
        for (const e of this.entries) {
            const note = e.note ? ` – ${e.note}` : '';
            const bmiStr = e.bmi ? ` (BMI: ${e.bmi.toFixed(1)})` : '';
            console.log(`  ${e.date}: ${e.weight.toFixed(1)} kg${bmiStr}${note}`);
        }
    }

    stats() {
        if (this.entries.length === 0) {
            console.log('No entries.');
            return;
        }
        const weights = this.entries.map(e => e.weight);
        const total = weights.length;
        const current = weights[weights.length - 1];
        const avg = weights.reduce((a, b) => a + b, 0) / total;
        const mn = Math.min(...weights);
        const mx = Math.max(...weights);
        const gain = this.preWeight ? current - this.preWeight : null;

        console.log('\n📊 Statistics:');
        console.log(`  Entries: ${total}`);
        console.log(`  Current: ${current.toFixed(1)} kg`);
        console.log(`  Average: ${avg.toFixed(1)} kg`);
        console.log(`  Min: ${mn.toFixed(1)} kg`);
        console.log(`  Max: ${mx.toFixed(1)} kg`);
        if (gain !== null) {
            console.log(`  Total gain: ${gain.toFixed(1)} kg`);
        }
    }

    showBMI() {
        if (this.entries.length === 0) {
            console.log('No entries.');
            return;
        }
        if (!this.height) {
            console.log('❌ Height not set.');
            return;
        }
        const current = this.entries[this.entries.length - 1].weight;
        const bmiVal = this.bmi(current);
        const cat = this.bmiCategory(bmiVal);
        console.log(`\n📐 Current BMI: ${bmiVal.toFixed(1)} (${cat})`);
        console.log('BMI Range: 18.5 - 24.9');
    }

    showGain() {
        if (this.entries.length === 0) {
            console.log('No entries.');
            return;
        }
        if (!this.preWeight) {
            console.log('❌ Pre‑pregnancy weight not set. Use: preweight <kg>');
            return;
        }
        const current = this.entries[this.entries.length - 1].weight;
        const first = this.entries[0].weight;
        const gain = current - this.preWeight;
        console.log('\n📈 Weight Gain:');
        console.log(`  Pre‑pregnancy: ${this.preWeight.toFixed(1)} kg`);
        console.log(`  First logged: ${first.toFixed(1)} kg`);
        console.log(`  Current: ${current.toFixed(1)} kg`);
        console.log(`  Total gain: ${gain.toFixed(1)} kg`);
    }

    trend(days = 5) {
        if (this.entries.length < days) {
            console.log(`Need at least ${days} entries for trend.`);
            return;
        }
        const recent = this.entries.slice(-days);
        const avg = recent.reduce((a, b) => a + b.weight, 0) / recent.length;
        console.log(`\n📈 Trend (last ${days} days):`);
        console.log(`  Average weight: ${avg.toFixed(1)} kg`);
        console.log('  Recent entries:');
        for (const e of recent.slice(-5)) {
            console.log(`    ${e.date}: ${e.weight.toFixed(1)} kg`);
        }
    }

    export(filename = 'weights.csv') {
        if (this.entries.length === 0) {
            console.log('No entries.');
            return;
        }
        let csv = 'Date,Weight (kg),BMI,Note\n';
        for (const e of this.entries) {
            const bmiStr = e.bmi ? e.bmi.toFixed(1) : '';
            csv += `${e.date},${e.weight.toFixed(1)},${bmiStr},${e.note}\n`;
        }
        fs.writeFileSync(filename, csv);
        console.log(`✅ Exported ${this.entries.length} entries to ${filename}`);
    }
}

program
    .command('height <cm>')
    .action((cm) => {
        const app = new PregnancyWeight();
        app.setHeight(parseFloat(cm));
    });

program
    .command('preweight <kg>')
    .action((kg) => {
        const app = new PregnancyWeight();
        app.setPreWeight(parseFloat(kg));
    });

program
    .command('add <weight>')
    .option('--date <date>', 'YYYY-MM-DD')
    .option('--note <note>', 'Optional note')
    .action((weight, options) => {
        const app = new PregnancyWeight();
        app.add(parseFloat(weight), options.date, options.note);
    });

program
    .command('list')
    .action(() => {
        const app = new PregnancyWeight();
        app.list();
    });

program
    .command('stats')
    .action(() => {
        const app = new PregnancyWeight();
        app.stats();
    });

program
    .command('bmi')
    .action(() => {
        const app = new PregnancyWeight();
        app.showBMI();
    });

program
    .command('gain')
    .action(() => {
        const app = new PregnancyWeight();
        app.showGain();
    });

program
    .command('trend')
    .option('--days <n>', 'Number of days', parseInt, 5)
    .action((options) => {
        const app = new PregnancyWeight();
        app.trend(options.days);
    });

program
    .command('export')
    .argument('[filename]', 'Output filename', 'weights.csv')
    .action((filename) => {
        const app = new PregnancyWeight();
        app.export(filename);
    });

program.parse(process.argv);

File: pregnancy_weight.rb
# pregnancy_weight.rb
#!/usr/bin/env ruby
require 'json'
require 'date'
require 'optparse'
require 'csv'

DATA_FILE = 'pregnancy_weight.json'

class PregnancyWeight
  attr_reader :height, :pre_weight, :entries

  def initialize
    @height = nil
    @pre_weight = nil
    @entries = []
    load
  end

  def load
    if File.exist?(DATA_FILE)
      data = JSON.parse(File.read(DATA_FILE))
      @height = data['height']
      @pre_weight = data['pre_weight']
      @entries = data['entries'] || []
    end
  end

  def save
    File.write(DATA_FILE, JSON.pretty_generate({
      height: @height,
      pre_weight: @pre_weight,
      entries: @entries
    }))
  end

  def set_height(cm)
    @height = cm
    save
    puts "✅ Height set to #{cm} cm"
  end

  def set_pre_weight(kg)
    @pre_weight = kg
    save
    puts "✅ Pre‑pregnancy weight set to #{kg} kg"
  end

  def bmi(weight)
    return nil unless @height
    h = @height / 100.0
    weight / (h * h)
  end

  def bmi_category(bmi_val)
    if bmi_val < 18.5
      "Underweight"
    elsif bmi_val < 25.0
      "Normal"
    elsif bmi_val < 30.0
      "Overweight"
    else
      "Obese"
    end
  end

  def add(weight, date_str = nil, note = '')
    unless @height
      puts "❌ Please set your height first: height <cm>"
      return
    end
    date_str ||= Date.today.to_s
    bmi_val = bmi(weight)
    entry = { 'date' => date_str, 'weight' => weight, 'note' => note, 'bmi' => bmi_val }
    @entries << entry
    @entries.sort_by! { |e| e['date'] }
    save
    puts "✅ Logged #{weight} kg on #{date_str}"
  end

  def list
    if @entries.empty?
      puts "No entries."
      return
    end
    puts "\n📋 Weight Entries:"
    @entries.each do |e|
      note = e['note'].empty? ? '' : " – #{e['note']}"
      bmi_str = e['bmi'] ? " (BMI: #{e['bmi'].round(1)})" : ''
      puts "  #{e['date']}: #{e['weight'].round(1)} kg#{bmi_str}#{note}"
    end
  end

  def stats
    if @entries.empty?
      puts "No entries."
      return
    end
    weights = @entries.map { |e| e['weight'] }
    total = weights.size
    current = weights.last
    avg = weights.sum / total
    mn = weights.min
    mx = weights.max
    gain = @pre_weight ? current - @pre_weight : nil

    puts "\n📊 Statistics:"
    puts "  Entries: #{total}"
    puts "  Current: #{current.round(1)} kg"
    puts "  Average: #{avg.round(1)} kg"
    puts "  Min: #{mn.round(1)} kg"
    puts "  Max: #{mx.round(1)} kg"
    puts "  Total gain: #{gain.round(1)} kg" if gain
  end

  def show_bmi
    if @entries.empty?
      puts "No entries."
      return
    end
    unless @height
      puts "❌ Height not set."
      return
    end
    current = @entries.last['weight']
    bmi_val = bmi(current)
    cat = bmi_category(bmi_val)
    puts "\n📐 Current BMI: #{bmi_val.round(1)} (#{cat})"
    puts "BMI Range: 18.5 - 24.9"
  end

  def show_gain
    if @entries.empty?
      puts "No entries."
      return
    end
    unless @pre_weight
      puts "❌ Pre‑pregnancy weight not set. Use: preweight <kg>"
      return
    end
    current = @entries.last['weight']
    first = @entries.first['weight']
    gain = current - @pre_weight
    puts "\n📈 Weight Gain:"
    puts "  Pre‑pregnancy: #{@pre_weight.round(1)} kg"
    puts "  First logged: #{first.round(1)} kg"
    puts "  Current: #{current.round(1)} kg"
    puts "  Total gain: #{gain.round(1)} kg"
  end

  def trend(days = 5)
    if @entries.size < days
      puts "Need at least #{days} entries for trend."
      return
    end
    recent = @entries.last(days)
    avg = recent.sum { |e| e['weight'] } / recent.size
    puts "\n📈 Trend (last #{days} days):"
    puts "  Average weight: #{avg.round(1)} kg"
    puts "  Recent entries:"
    recent.last(5).each do |e|
      puts "    #{e['date']}: #{e['weight'].round(1)} kg"
    end
  end

  def export(filename = 'weights.csv')
    if @entries.empty?
      puts "No entries."
      return
    end
    CSV.open(filename, 'w') do |csv|
      csv << ['Date', 'Weight (kg)', 'BMI', 'Note']
      @entries.each do |e|
        bmi_str = e['bmi'] ? e['bmi'].round(1).to_s : ''
        csv << [e['date'], e['weight'].round(1), bmi_str, e['note']]
      end
    end
    puts "✅ Exported #{@entries.size} entries to #{filename}"
  end
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: pregnancy_weight.rb <command> [options]"
  opts.on("height CM", Float, "Set height") { |v| options[:height] = v }
  opts.on("preweight KG", Float, "Set pre‑pregnancy weight") { |v| options[:preweight] = v }
  opts.on("add WEIGHT", Float, "Log weight") { |v| options[:add] = v }
  opts.on("--date DATE", "YYYY-MM-DD") { |v| options[:date] = v }
  opts.on("--note TEXT", "Optional note") { |v| options[:note] = v }
  opts.on("list", "List entries") { options[:list] = true }
  opts.on("stats", "Show statistics") { options[:stats] = true }
  opts.on("bmi", "Show BMI") { options[:bmi] = true }
  opts.on("gain", "Show weight gain") { options[:gain] = true }
  opts.on("trend", "Show trend") { options[:trend] = true }
  opts.on("--days N", Integer, "Days for trend") { |v| options[:trend_days] = v }
  opts.on("export", "Export to CSV") { options[:export] = true }
  opts.on("--filename FILE", "Output filename") { |v| options[:filename] = v }
end.parse!

app = PregnancyWeight.new

if options[:height]
  app.set_height(options[:height])
elsif options[:preweight]
  app.set_pre_weight(options[:preweight])
elsif options[:add]
  app.add(options[:add], options[:date], options[:note] || '')
elsif options[:list]
  app.list
elsif options[:stats]
  app.stats
elsif options[:bmi]
  app.show_bmi
elsif options[:gain]
  app.show_gain
elsif options[:trend]
  app.trend(options[:trend_days] || 5)
elsif options[:export]
  app.export(options[:filename] || 'weights.csv')
else
  puts "Usage: pregnancy_weight.rb <command> [options]"
end
